### ISSUE TITLE
Can RewardCriterion backward() get the correct gradient?

### ISSUE THREAD

**Issue by lujiaying** [ruotianluo/self-critical.pytorch#64]
Hi Ruotian,

I read your notes in [Zhihu](https://zhuanlan.zhihu.com/p/24879932) as well. Your note helps me a lot.
Thank you so much.

One concern for me is, as formulation (9) in "Self-critical Sequence Training for Image Captioning" indicates the way to calculate the gradient, can the default Pytorch backward() function get the correct gradient?  The formulation (9) directly calculates the gradient of `S_t` which is the input of softmax layer, while in my mind the default Pytorch backward() is for the gradient of `the output of softmax layer`.

Cheers,
Jiaying

**Comment by ruotianluo** [comment]
The input of the rewardcriterion is the output of logsoftmax. This loss itself doesn't have theoretical meaning; the loss is defined in this way so that if you take backward on the loss, the gradient got is exactly equation 9.

**Comment by lujiaying** [comment]
Thank you for your reply.

![image](https://user-images.githubusercontent.com/5207337/47599366-62f3cc80-d961-11e8-9d51-cf69c866d891.png)
For this gradient(differentiating L(theta) with log(p_theta)), will the -(r-b) updates every index in the dict-dimension vector or only the index of sampled word in the dict-dimension vector?

P.S. 为了防止英文说不清楚。
想请教下，L(theta)对 logp_theta求导的话，-(r-b)会更新到softmax输出向量的每一维度，还是类似于cross-entropy只更新target_index那一维度呢？
